### ISSUE TITLE
[no-unused-vars] false positive when class property has the same name as import

### ISSUE THREAD

**Issue by kmaraz** [typescript-eslint/typescript-eslint#2994]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

 https://github.com/kmaraz/no-unused-vars-issue
I'll show you the issue in the pictures:
![image](https://user-images.githubusercontent.com/6079654/106623923-667d5180-6575-11eb-9e74-d6442c2157e2.png)
This works as the class property `thisWorks` has different name from the `Test`.
![image](https://user-images.githubusercontent.com/6079654/106624010-7eed6c00-6575-11eb-8b28-fec47082e488.png)
This does not work, as the class property `Test` has the same name as `Test` import.
Hope it helps! :)
**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.14.2` |
| `@typescript-eslint/parser`        | `4.14.2` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.18.0` |
| `node`                             | `14.15.0` |

**Comment by bradzacher** [comment]
I believe that the bug is here:

https://github.com/typescript-eslint/typescript-eslint/blob/7695ef318f1cc8688acaabf4f2730769622f083f/packages/scope-manager/src/referencer/ClassVisitor.ts#L298

This is adding a reference to the `Test` value in the current scope.
However the current scope is the function scope - and the function scope has a `Test` variable defined (the parameter).
This should be referencing the value from the parent scope instead.

**Comment by splincode** [comment]
I same this issue:
https://github.com/Angular-RU/angular-ru-sdk/pull/361

**Comment by bradzacher** [comment]
This is a pretty complex problem to solve for all cases.
Name shadowing sucks to deal with.

On the one hand - one can naively fix this by placing the reference in the parent scope.
However this would then cause this case to break:
```ts
import { Test } from './Test';

declare function deco(..._param: any): any;

export class Clazz {
  @deco
  method<Test>(Test: Test) {}
}
```
I've opened an issue with TypeScript to clarify how it's supposed to work.

**Comment by vegerot** [comment]
> I've opened an issue with TypeScript to clarify how it's supposed to work.

@bradzacher link to this issue?

**Comment by bradzacher** [comment]
https://github.com/microsoft/TypeScript/issues/42679

**Comment by vegerot** [comment]
🙂 thanks

**Comment by armano2** [comment]
i did small investigation to this issue 

```ts
import { Test } from './zzzz';

function Component(test: any) {
  return function (constructor: any) {
    console.log(test)
    Object.seal(constructor);
  }
}

@Component({})
export class Index {
  constructor(public Test: Test) { // Variable has the same name as type declaration

  }
}

```

```json
{
  "compileOnSave": false,
  "compilerOptions": {
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
  }
}
```

issue disappears when i remove `emitDecoratorMetadata`, but i'm unable to reproduce it in playground :(

**Comment by Woebin** [comment]
There was a response to the TypeScript issue logged by @bradzacher two weeks ago, see https://github.com/microsoft/TypeScript/issues/42679#issuecomment-1917856418 - the response states that the described behavior is expected.

Commenting to point this out as I'm seeing the same linting error as described here, using @typescript-eslint/eslint-plugin@6.7.5 - since this issue is still open, I'm guessing it's not considered resolved in a later version of the plugin either?
